File: src/main/python/com/example/tutorial/dsa/medium/simulation/rate_limiter_controller.py
"""
RateLimiterController
----------------------------------
This program implements rate limiting algorithms for API throttling.
The core problem solved here is Logger Rate Limiter (LeetCode #359).

Problem Statement:
    Design a logger system that receives a stream of messages and timestamps,
    and returns true if the message should be printed (not printed in the last 10 seconds).

Real UseCase:
    In a credit card offers system:
    - Rate limit API requests per user/merchant
    - Throttle duplicate transaction alerts
    - Control notification frequency to users

Company Tags: Google, Amazon, Facebook

See: https://leetcode.com/problems/logger-rate-limiter/
"""

import sys
import os
from collections import deque
from typing import Dict

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "..", ".."))

from com.example.tutorial.common.utils.sample_data_loader import load_offers


class Logger:
    """
    Logger class implementing rate limiting for messages.

    LOGIC (HashMap with Timestamp Tracking):
        1. Store last printed timestamp for each message
        2. On new request, check if 10+ seconds have passed
        3. If yes, update timestamp and return true (print)
        4. If no, return false (skip)

    Time Complexity: O(1) per operation.
    Space Complexity: O(n) where n = unique messages.
    """

    def __init__(self):
        """Initialize the logger."""
        self.message_timestamps: Dict[str, int] = {}

    def should_print_message(self, timestamp: int, message: str) -> bool:
        """
        Returns true if the message should be printed (not printed in last 10 seconds).

        Args:
            timestamp: Current timestamp in seconds.
            message: The message to log.

        Returns:
            True if should print, False otherwise.
        """
        if message not in self.message_timestamps or \
           timestamp - self.message_timestamps[message] >= 10:
            self.message_timestamps[message] = timestamp
            return True
        return False


class SlidingWindowRateLimiter:
    """
    Sliding Window Rate Limiter using timestamp queue.

    LOGIC (Queue of Timestamps):
        1. For each user, maintain a queue of request timestamps
        2. On new request, remove expired timestamps (outside window)
        3. If queue size < limit, allow request and add timestamp
        4. Otherwise, reject request

    Time Complexity: O(w) where w = window size (cleanup).
    Space Complexity: O(n x w) where n = users, w = max requests per window.
    """

    def __init__(self, max_requests: int, window_seconds: int):
        """
        Initialize the rate limiter.

        Args:
            max_requests: Maximum requests allowed per window.
            window_seconds: Size of the sliding window in seconds.
        """
        self.max_requests = max_requests
        self.window_seconds = window_seconds
        self.user_requests: Dict[str, deque] = {}

    def allow_request(self, user_id: str, timestamp: int) -> bool:
        """
        Checks if request is allowed and records it if so.

        Args:
            user_id: The user making the request.
            timestamp: Current timestamp in seconds.

        Returns:
            True if request is allowed.
        """
        if user_id not in self.user_requests:
            self.user_requests[user_id] = deque()
        requests = self.user_requests[user_id]

        # Remove expired timestamps.
        while requests and timestamp - requests[0] >= self.window_seconds:
            requests.popleft()

        # Check if under limit.
        if len(requests) < self.max_requests:
            requests.append(timestamp)
            return True
        return False


class TokenBucket:
    """
    Token Bucket Rate Limiter.

    LOGIC (Token Bucket Algorithm):
        1. Bucket starts with max tokens
        2. Tokens refill at constant rate up to max
        3. Each request consumes tokens
        4. Request allowed only if enough tokens available

    Time Complexity: O(1) per operation.
    Space Complexity: O(1).
    """

    def __init__(self, max_tokens: int, refill_rate: float):
        """
        Initialize the token bucket.

        Args:
            max_tokens: Maximum capacity of the bucket.
            refill_rate: Tokens added per second.
        """
        self.max_tokens = max_tokens
        self.refill_rate = refill_rate
        self.tokens = float(max_tokens)
        self.last_refill_time = 0

    def refill(self, current_time: int) -> None:
        """
        Refills tokens based on elapsed time.

        Args:
            current_time: Current timestamp in seconds.
        """
        elapsed = current_time - self.last_refill_time
        self.tokens = min(self.max_tokens, self.tokens + elapsed * self.refill_rate)
        self.last_refill_time = current_time

    def try_consume(self, count: int) -> bool:
        """
        Attempts to consume tokens.

        Args:
            count: Number of tokens to consume.

        Returns:
            True if tokens were consumed successfully.
        """
        if self.tokens >= count:
            self.tokens -= count
            return True
        return False

    def get_tokens(self) -> float:
        """Returns current token count."""
        return self.tokens


def main():
    """Main function to demonstrate the RateLimiterController."""
    print("=== RateLimiterController: Rate Limiting Demo ===\n")

    # Load credit card offers from sample data
    offers = load_offers()
    print(f"Loaded {len(offers)} offers from sample data.\n")

    # Demonstrate Logger Rate Limiter
    print("--- Logger Rate Limiter (10 second window) ---\n")
    logger = Logger()

    messages = ["foo", "bar", "foo", "bar", "foo", "foo"]
    timestamps = [1, 2, 3, 8, 10, 11]

    for i in range(len(messages)):
        result = logger.should_print_message(timestamps[i], messages[i])
        status = "PRINT" if result else "SKIP"
        print(f"timestamp={timestamps[i]}, message=\"{messages[i]}\" -> {status}")

    # Demonstrate Sliding Window Rate Limiter
    print("\n--- Sliding Window Rate Limiter (3 requests per 5 seconds) ---\n")
    rate_limiter = SlidingWindowRateLimiter(3, 5)

    user_id = "user123"
    request_times = [1, 2, 3, 4, 5, 6, 7, 10, 11, 12]

    for time in request_times:
        allowed = rate_limiter.allow_request(user_id, time)
        status = "ALLOWED" if allowed else "RATE LIMITED"
        print(f"time={time}, user={user_id} -> {status}")

    # Demonstrate Token Bucket
    print("\n--- Token Bucket Rate Limiter (5 tokens, 1 token/sec refill) ---\n")
    token_bucket = TokenBucket(5, 1.0)

    for time in range(0, 16, 2):
        token_bucket.refill(time)
        allowed = token_bucket.try_consume(1)
        status = "ALLOWED" if allowed else "RATE LIMITED"
        print(f"time={time} -> {status} (tokens: {token_bucket.get_tokens():.1f})")


if __name__ == "__main__":
    main()

File: src/main/java/com/example/tutorial/dsa/medium/simulation/RateLimiterController.java
package com.example.tutorial.dsa.medium.simulation;

import com.example.tutorial.common.datamodel.Offer;
import com.example.tutorial.common.utils.SampleDataLoader;
import org.springframework.boot.CommandLineRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.stereotype.Component;

import java.util.*;

/**
 * RateLimiterController
 * ----------------------------------
 * <p>This program implements rate limiting algorithms for API throttling.
 * The core problem solved here is Logger Rate Limiter (LeetCode #359).
 *
 * <p><b>Problem Statement:</b>
 * Design a logger system that receives a stream of messages and timestamps,
 * and returns true if the message should be printed (not printed in the last 10 seconds).
 *
 * <p><b>Real UseCase:</b> In a credit card offers system:
 * <ul>
 *   <li>Rate limit API requests per user/merchant</li>
 *   <li>Throttle duplicate transaction alerts</li>
 *   <li>Control notification frequency to users</li>
 * </ul>
 *
 * <p><b>Company Tags:</b> Google, Amazon, Facebook
 *
 * @see <a href="https://leetcode.com/problems/logger-rate-limiter/">LeetCode 359 - Logger Rate Limiter</a>
 */
@Component
public class RateLimiterController implements CommandLineRunner {
  public static void main(String[] args) {
    SpringApplication.run(RateLimiterController.class, args);
  }

  @Override
  public void run(String... args) throws Exception {
    System.out.println("=== RateLimiterController: Rate Limiting Demo ===\n");

    // Load credit card offers from sample data
    List<Offer> offers = SampleDataLoader.OFFERS_DTO.get();
    System.out.println("Loaded " + offers.size() + " offers from sample data.\n");

    // Demonstrate Logger Rate Limiter
    System.out.println("--- Logger Rate Limiter (10 second window) ---\n");
    Logger logger = new Logger();

    String[] messages = {"foo", "bar", "foo", "bar", "foo", "foo"};
    int[] timestamps = {1, 2, 3, 8, 10, 11};

    for (int i = 0; i < messages.length; i++) {
      boolean result = logger.shouldPrintMessage(timestamps[i], messages[i]);
      System.out.println("timestamp=" + timestamps[i] + ", message=\"" + messages[i]
          + "\" → " + (result ? "PRINT" : "SKIP"));
    }

    // Demonstrate Sliding Window Rate Limiter
    System.out.println("\n--- Sliding Window Rate Limiter (3 requests per 5 seconds) ---\n");
    SlidingWindowRateLimiter rateLimiter = new SlidingWindowRateLimiter(3, 5);

    String userId = "user123";
    int[] requestTimes = {1, 2, 3, 4, 5, 6, 7, 10, 11, 12};

    for (int time : requestTimes) {
      boolean allowed = rateLimiter.allowRequest(userId, time);
      System.out.println("time=" + time + ", user=" + userId
          + " → " + (allowed ? "ALLOWED" : "RATE LIMITED"));
    }

    // Demonstrate Token Bucket
    System.out.println("\n--- Token Bucket Rate Limiter (5 tokens, 1 token/sec refill) ---\n");
    TokenBucket tokenBucket = new TokenBucket(5, 1.0);

    for (int time = 0; time <= 15; time += 2) {
      tokenBucket.refill(time);
      boolean allowed = tokenBucket.tryConsume(1);
      System.out.println("time=" + time + " → " + (allowed ? "ALLOWED" : "RATE LIMITED")
          + " (tokens: " + String.format("%.1f", tokenBucket.getTokens()) + ")");
    }
  }

  /**
   * Logger class implementing rate limiting for messages.
   *
   * <p><b>LOGIC (HashMap with Timestamp Tracking):</b>
   * <ol>
   *   <li>Store last printed timestamp for each message</li>
   *   <li>On new request, check if 10+ seconds have passed</li>
   *   <li>If yes, update timestamp and return true (print)</li>
   *   <li>If no, return false (skip)</li>
   * </ol>
   *
   * <p><b>Time Complexity: O(1)</b> per operation.
   * <p><b>Space Complexity: O(n)</b> where n = unique messages.
   */
  public static class Logger {
    private final Map<String, Integer> messageTimestamps;

    public Logger() {
      messageTimestamps = new HashMap<>();
    }

    /**
     * Returns true if the message should be printed (not printed in last 10 seconds).
     *
     * @param timestamp current timestamp in seconds
     * @param message the message to log
     * @return true if should print, false otherwise
     */
    public boolean shouldPrintMessage(int timestamp, String message) {
      if (!messageTimestamps.containsKey(message)
          || timestamp - messageTimestamps.get(message) >= 10) {
        messageTimestamps.put(message, timestamp);
        return true;
      }
      return false;
    }
  }

  /**
   * Sliding Window Rate Limiter using timestamp queue.
   *
   * <p><b>LOGIC (Queue of Timestamps):</b>
   * <ol>
   *   <li>For each user, maintain a queue of request timestamps</li>
   *   <li>On new request, remove expired timestamps (outside window)</li>
   *   <li>If queue size < limit, allow request and add timestamp</li>
   *   <li>Otherwise, reject request</li>
   * </ol>
   *
   * <p><b>Time Complexity: O(w)</b> where w = window size (cleanup).
   * <p><b>Space Complexity: O(n × w)</b> where n = users, w = max requests per window.
   */
  public static class SlidingWindowRateLimiter {
    private final int maxRequests;
    private final int windowSeconds;
    private final Map<String, Deque<Integer>> userRequests;

    public SlidingWindowRateLimiter(int maxRequests, int windowSeconds) {
      this.maxRequests = maxRequests;
      this.windowSeconds = windowSeconds;
      this.userRequests = new HashMap<>();
    }

    /**
     * Checks if request is allowed and records it if so.
     *
     * @param userId the user making the request
     * @param timestamp current timestamp in seconds
     * @return true if request is allowed
     */
    public boolean allowRequest(String userId, int timestamp) {
      userRequests.putIfAbsent(userId, new LinkedList<>());
      Deque<Integer> requests = userRequests.get(userId);

      // Remove expired timestamps.
      while (!requests.isEmpty() && timestamp - requests.peekFirst() >= windowSeconds) {
        requests.pollFirst();
      }

      // Check if under limit.
      if (requests.size() < maxRequests) {
        requests.addLast(timestamp);
        return true;
      }
      return false;
    }
  }

  /**
   * Token Bucket Rate Limiter.
   *
   * <p><b>LOGIC (Token Bucket Algorithm):</b>
   * <ol>
   *   <li>Bucket starts with max tokens</li>
   *   <li>Tokens refill at constant rate up to max</li>
   *   <li>Each request consumes tokens</li>
   *   <li>Request allowed only if enough tokens available</li>
   * </ol>
   *
   * <p><b>Time Complexity: O(1)</b> per operation.
   * <p><b>Space Complexity: O(1)</b>.
   */
  public static class TokenBucket {
    private final int maxTokens;
    private final double refillRate;  // Tokens per second.
    private double tokens;
    private int lastRefillTime;

    public TokenBucket(int maxTokens, double refillRate) {
      this.maxTokens = maxTokens;
      this.refillRate = refillRate;
      this.tokens = maxTokens;
      this.lastRefillTime = 0;
    }

    /**
     * Refills tokens based on elapsed time.
     *
     * @param currentTime current timestamp in seconds
     */
    public void refill(int currentTime) {
      int elapsed = currentTime - lastRefillTime;
      tokens = Math.min(maxTokens, tokens + elapsed * refillRate);
      lastRefillTime = currentTime;
    }

    /**
     * Attempts to consume tokens.
     *
     * @param count number of tokens to consume
     * @return true if tokens were consumed successfully
     */
    public boolean tryConsume(int count) {
      if (tokens >= count) {
        tokens -= count;
        return true;
      }
      return false;
    }

    public double getTokens() {
      return tokens;
    }
  }
}
